### DISCUSSION TITLE
Shadow DOM changes ignored

### DISCUSSION THREAD

**Discussion by luttje** [livewire/livewire#5021] [3. Feature ideas]
Hi there,

First of all: love Livewire, good stuff, thanks for all your work!

*For clarity: I have updated this post multiple times (and removed some comments of my own) as I discovered more about my own bugs and potential bugs in Livewire*

## Goal
I'm trying to "scope css" and prevent styles from leaking out of components into the rest of the page. I thought I'd try using Shadow DOM for this which works fine on it's own. Though I was having some trouble getting it to work in combination with Livewire. In the end I managed to get it to work once with a workaround:
![image](https://user-images.githubusercontent.com/2738114/172716251-f625c52c-2e65-4176-a470-997f6113f8f1.png)
*Image from [the readme of this proof-of-concept package](https://github.com/luttje/laravel-scoped-components)*

## Problems/Bugs
**Sadly neither Livewire nor AlpineJS work properly when using the Shadow DOM ([attachShadow](https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow))**. They do not detect children in the shadow dom, and will not by themselves attach their events to wired elements.

I've created [a repo here](https://github.com/luttje/livewire-shadow-dom-bug-demo) so it's easy to reproduce the problem. It also contains [a hacky workaround](https://github.com/luttje/livewire-shadow-dom-bug-demo/blob/main/resources/views/js-workarounds.blade.php) to "fix" this problem, but that only works on first load and breaks after clicking once.

The biggest problems I'm running into are:
1. Elements in the Shadow DOM are not seen by Livewire. Their wire directives do not work.
2. Changes in a &lt;template&gt;'s .content aren't updated by Livewire, which I believe is due to the use of isEqualNode (https://jsfiddle.net/vryfm2sh/)
3. <s>After I update the DOM using Javascript, I can't find a reliable way to have Livewire attach it's event handler to my wire:click button inside &lt;template&gt; or the shadow dom.</s> Just use `component.__livewire.tearDown()` and then `.__livewire.initialize()` to re-attach wire directives.
4. Is there a way to re-attach wire directives to elements that are missing it, without calling `.__livewire.tearDown()` and then `.__livewire.initialize()`? Because doing so [reevaluates script tags](https://github.com/livewire/livewire/blob/3ce704ac3314fc8713f710ee1c4e865846569c19/js/node_initializer.js#L11), which is not what I'm looking for.

## Fix(es)
I made a fork with multiple branches. The numbers below match the problems above:
1. This branch also includes shadowRoot content when walking through the dom: https://github.com/livewire/livewire/compare/master...luttje:add-shadow-dom-tests
2. Here's a branch with a suggestion on how to fix the template content problem. It isn't very elegant, but it's something: https://github.com/livewire/livewire/compare/master...luttje:add-template-update-test
3. I played a bit with my originally problematic code and found that after simplifying it, I could simply tearDown and initialize the component to restore wire directives. I found that out while writing a test. For reference here is that test (may be useful): https://github.com/livewire/livewire/compare/master...luttje:add-dynamic-change-tests
4. No clean fix for this yet. Setting `window.Livewire.components.initialRenderIsFinished = false` before calling a components initialize works, but feels hacky.

## Questions
- <s>Is this a bug or am I doing something wrong?</s> I was doing something wrong, and there is some bugs with template and shadow dom elements
- Could I get some C&C on my fork? I find it challenging to find the "right way" to approach these fixes, without some input from others
- Can I create some PR's for the fork branches I have created?

Thanks for your time!

**Comment by joshhanley** [comment]
@luttje Livewire doesn't currently support ShadowDOM, as such I wouldn't call it a bug, so I've update the category to a feature request 🙂

All PR's are welcome though, so feel free to submit one if you have it working 🙂 What's C&C (other than chips 😆)?

**Reply by luttje** [reply]
Hey and thanks for your reply! Sorry for the ambiguous acronym, I meant "Comments and Criticism" (though those chips are appreciated as well 😋)

I'll have a look at my forks later today or this weekend and see if I need to make changes before I make PR's.

I do have a question though (since the forks work in this way): Is the use of 
```js
const component = template.closest('[wire\\3A id]')
component.__livewire.tearDown()
compoment.__livewire.initialize()
```


to re-attach wire directives intended?

Since shadow DOM only exists when JavaScript places objects in it I'm looking for a way to refresh a livewire component on the JavaScript side.

Thanks and have a nice day!

**Reply by joshhanley** [reply]
@luttje ah haha that makes sense 😆

Ok cool. 

Hmm I'm not sure what you mean, can you link me to where that is in the code base? Then I can have a look 🙂

**Reply by luttje** [reply]
Thanks! This is where I use it after putting something in Shadow DOM: [link](https://github.com/livewire/livewire/compare/master...luttje:add-shadow-dom-tests#diff-45b972fd682ecdfc84d23482e22762d7ed50bcd28112107a8de3de25fabc9a45R19)

It works in that fork because I [made the dom walker also look for shadow root](https://github.com/livewire/livewire/compare/master...luttje:add-shadow-dom-tests#diff-c8d9b3a62815990c436a8a71e94a54dfb1c7eb195bb6b520dd0a64563439b9f6R7).

**Reply by joshhanley** [reply]
@luttje ah right ok, I haven't got the time at the moment to dig in, but if I get some time next week, I'll have a closer look at it for you 🙂 Can I suggest submitting the PR anyway (with the above code included) then we can review it through the PR process?
